FUNCTION_TO_MODULE = {
        'get_total_memory_usage': 'system',
        'display_top_memory_usage': 'system',
        'get_available_program_names_in_active_environment': 'system',
        'is_program_exists': 'system',
        'check_h5py_module': 'system',

        'rev_comp': 'sequences',
        'rev_comp_gene_calls_dict': 'sequences',
        'get_GC_content_for_sequence': 'sequences',
        'get_synonymous_and_non_synonymous_potential': 'sequences',
        'get_list_of_AAs_for_gene_call': 'sequences',
        'get_list_of_codons_for_gene_call': 'sequences',
        'get_translated_sequence_for_gene_call': 'sequences',
        'translate': 'sequences',
        'get_most_likely_translation_frame': 'sequences',
        'get_codon_order_to_nt_positions_dict': 'sequences',
        'nt_seq_to_nt_num_array': 'sequences',
        'nt_seq_to_RC_nt_num_array': 'sequences',
        'nt_seq_to_codon_num_array': 'sequences',
        'nt_seq_to_RC_codon_num_array': 'sequences',
        '_nt_seq_to_codon_num_array': 'sequences',
        'is_amino_acid_functionally_conserved': 'sequences',
        'convert_sequence_indexing': 'sequences',
        'summarize_alignment': 'sequences',
        'restore_alignment': 'sequences',

        'serialize_args': 'commandline',
        'format_cmdline': 'commandline',
        'run_command': 'commandline',
        'start_command': 'commandline',
        'run_command_STDIN': 'commandline',
        'get_command_output_from_shell': 'commandline',
        'get_cmd_line': 'commandline',
        'get_gene_caller_ids_from_args': 'commandline',

        'gzip_compress_file': 'files',
        'gzip_decompress_file': 'files',
        'tar_extract_file': 'files',
        'store_array_as_TAB_delimited_file': 'files',
        'store_dataframe_as_TAB_delimited_file': 'files',
        'store_dict_as_TAB_delimited_file': 'files',
        'get_column_data_from_TAB_delim_file': 'files',
        'get_columns_of_TAB_delim_file': 'files',
        'is_all_columns_present_in_TAB_delim_file': 'files',
        'transpose_tab_delimited_file': 'files',
        'get_vectors_from_TAB_delim_matrix': 'files',
        'get_TAB_delimited_file_as_dictionary': 'files',
        'get_BLAST_tabular_output_as_dict': 'files',
        'get_bams_and_profiles_txt_as_data': 'files',
        'get_samples_txt_file_as_dict': 'files',
        'get_primers_txt_file_as_dict': 'files',
        'get_groups_txt_file_as_dict': 'files',
        'get_yaml_as_dict': 'files',
        'get_chunk': 'files',
        'concatenate_files': 'files',
        'get_file_md5': 'files',
        'human_readable_file_size': 'files',

        'split_fasta': 'fasta',
        'get_num_sequences_in_fasta': 'fasta',
        'get_all_ids_from_fasta': 'fasta',
        'check_fasta_id_formatting': 'fasta',
        'check_fasta_id_uniqueness': 'fasta',
        'get_read_lengths_from_fasta': 'fasta',
        'remove_sequences_with_only_gaps_from_fasta': 'fasta',
        'get_FASTA_file_as_dictionary': 'fasta',
        'get_GC_content_for_FASTA_entries': 'fasta',
        'unique_FASTA_file': 'fasta',
        'store_dict_as_FASTA_file': 'fasta',
        'export_sequences_from_contigs_db': 'fasta',
        'create_fasta_dir_from_sequence_sources': 'fasta',

        'get_db_type_and_variant': 'database',
        'get_db_type': 'database',
        'get_db_variant': 'database',
        'get_required_version_for_db': 'database',
        'get_all_sample_names_from_the_database': 'database',
        'get_all_item_names_from_the_database': 'database',
        'get_genes_database_path_for_bin': 'database',

        'is_contigs_db': 'dbcheck',
        'is_trnaseq_db': 'dbcheck',
        'is_pan_or_profile_db': 'dbcheck',
        'is_profile_db': 'dbcheck',
        'is_structure_db': 'dbcheck',
        'is_blank_profile': 'dbcheck',
        'is_pan_db': 'dbcheck',
        'is_genome_storage': 'dbcheck',
        'is_genes_db': 'dbcheck',
        'is_kegg_modules_db': 'dbcheck',
        'is_profile_db_merged': 'dbcheck',
        'is_profile_db_and_contigs_db_compatible': 'dbcheck',
        'is_structure_db_and_contigs_db_compatible': 'dbcheck',
        'is_pan_db_and_genomes_storage_db_compatible': 'dbcheck',

        'check_sample_id': 'validation',
        'check_collection_name': 'validation',
        'is_this_name_OK_for_database': 'validation',
        'check_contig_names': 'validation',
        'check_misc_data_keys_for_format': 'validation',
        'is_gene_caller_id': 'validation',
        'is_gene_sequence_clean': 'validation',
        'is_ascii_only': 'validation',

        'get_port_num': 'network',
        'get_next_available_port_num': 'network',
        'is_port_in_use': 'network',
        'download_file': 'network',
        'get_remote_file_content': 'network',
        'get_anvio_news': 'network',
        'download_protein_structure': 'network',
        'open_url_in_browser': 'network',

        'get_predicted_type_of_items_in_a_dict': 'misc',
        'HTMLColorToRGB': 'misc',
        'get_random_colors_dict': 'misc',
        'get_ordinal_from_integer': 'misc',
        'get_f_string_evaluated_by_dict': 'misc',
        'get_time_to_date': 'misc',
        'get_hash_for_list': 'misc',
        'get_filtered_dict': 'misc',

        'convert_numpy_array_to_binary_blob': 'algorithms',
        'convert_binary_blob_to_numpy_array': 'algorithms',
        'add_to_2D_numeric_array': 'algorithms',
        'apply_and_concat': 'algorithms',
        'get_indices_for_outlier_values': 'algorithms',
        'get_list_of_outliers': 'algorithms',
        'get_stretches_for_numbers_list': 'algorithms',
        'merge_stretches': 'algorithms',
        'get_constant_value_blocks': 'algorithms',
        'find_value_index': 'algorithms',
        'get_N50': 'algorithms',
        'split_by_delim_not_within_parens': 'algorithms',
        'get_split_start_stops': 'algorithms',
        'get_split_start_stops_with_gene_calls': 'algorithms',
        'get_split_start_stops_without_gene_calls': 'algorithms',

        'anvio_hmm_target_term_to_alphabet_and_context': 'hmm',
        'get_pruned_HMM_hits_dict': 'hmm',
        'get_HMM_sources_dictionary': 'hmm',
        'get_attribute_from_hmm_file': 'hmm',
        'sanity_check_hmm_model': 'hmm',
        'sanity_check_pfam_accessions': 'hmm',

        'run_functional_enrichment_stats': 'statistics',
        'get_enriched_groups': 'statistics',
        'get_required_packages_for_enrichment_test': 'statistics',
        'check_R_packages_are_installed': 'statistics',
        'get_values_of_gene_level_coverage_stats_as_dict': 'statistics',

        'get_names_order_from_newick_tree': 'phylogenetics',
        'gen_NEXUS_format_partition_file_for_phylogenomics': 'phylogenetics',

        'ununique_BLAST_tabular_output': 'anviohelp',
        'convert_SSM_to_single_accession': 'anviohelp',
        'get_default_gene_caller': 'anviohelp',
        'get_split_and_contig_names_of_interest': 'anviohelp',
        'get_contigs_splits_dict': 'anviohelp',
        'get_variabile_item_frequencies': 'anviohelp',
        'get_consensus_and_departure_data': 'anviohelp',
        'get_bin_name_from_item_name': 'anviohelp',
        'get_contig_name_to_splits_dict': 'anviohelp',
        'get_variability_table_engine_type': 'anviohelp',

        'gen_gexf_network_file': 'visualization',
        'run_selenium_and_export_svg': 'visualization',

        'compare_times': 'debug',
        'is_all_npm_packages_installed': 'debug',
}
