A fixation index matrix is what it sounds like (a matrix of fixation indices) and is generated by %(anvi-gen-fixation-index-matrix)s. 

This is a distance matrix where each column represnts a metagenome/sample and each row represents a metagenome/sample, so each cell represents the fixation index between two samples. The fixation index is a number from 0 (maximum similiarity) to 1 (maximum distance). 

This is a form of %(view-data)s, so it can be provided to %(anvi-matrix-to-newick)s as is done in the [Infant Gut tutorial](https://merenlab.org/tutorials/infant-gut/#measuring-distances-between-metagenomes-with-fst). 

For example, here is the fixation index matrix based off of a few random genes (1, 2, 3, 5, 6, 7, 8, 9, 21, 32, 35, 56, 567) in the infant gut tutorial: 

               DAY_15A               DAY_15B                DAY_16                 DAY_17A                 DAY_17B               DAY_18                  DAY_19                  DAY_22A                 DAY_22B                DAY_23                  DAY_24
    DAY_15A    0.0                   0.04918776635439459    0.11098963862572608    0.047199977957045336    0.0                   0.08439930158545839     0.06141920095408482     0.09624218229853498     0.0                    0.04711838006230529     0.08709060259498314
    DAY_15B    0.04918776635439459   0.0                    0.07095764349628797    0.04742239843190177     0.04918776635439459   0.058754132113538526    0.012860485049109083    0.056111593790529324    0.04918776635439459    0.012757678005705375    0.05031628777187558

    ...
