A KEGG KGML pathway walker analyzes a KGML file representation of a KEGG pathway to identify chains of compounds linked by reactions. This walker object is an instance of `anvio.kgmlnetworkops.KGMLNetworkWalker`. A walker's `get_chains` method returns `anvio.kgmlnetworkops.Chain` objects that show routes by which a compound is produced or consumed in a pathway map.

## Pathway maps

A walker navigates a KEGG pathway map specified by an ID, e.g., '00010' for 'Glycolysis / Gluconeogenesis'. The ID is stored in the `kegg_pathway_number` attribute of the walker. The map must have a KEGG reaction (RN) type KGML file available. Pathways with an RN type file are in categories 1.0 - 1.11 of KEGG's map classification: https://www.genome.jp/kegg/pathway.html. Given a pathway, the walker loads corresponding KGML files in an anvi'o KEGG installation assumed to be at the default location, with information on the installation stored in the walker's `kegg_context` attribute. Pathway maps can also be associated with KO (KEGG Ortholog) and EC (Enzyme Commission) type KGML files. RN type files contain information on all map reactions, including reactions catalyzed by enzymes represented by KOs and EC numbers, which are also contained in the KO and EC type KGML files, respectively, as well as nonenzymatic reactions, which are only contained in the RN type file. Nonenzymatic reactions are identified in the walker's `pathway_nonenzymatic_kgml_reaction_ids` attribute.

## Genomic routes

A walker can find metabolic routes genomically encoded by an organism or community. A KGML file indicates KOs capable of catalyzing reactions, and the walker seeks these KOs among the KO gene annotations stored in the %(contigs-db)s for the organism or community. The walker is initialized with the contigs database file path, as recorded in the walker's attribute, `contigs_db_path`. It is currently required that the contigs database be annotated with a %(reaction-network)s, which can be generated by running %(anvi-reaction-network)s. The walker's `network` attribute references the network object, an instance of `anvio.reactionnetwork.GenomicNetwork`. The walker can alternatively compare the pathway to a reaction network in the absence of an associated contigs database. For example, the metabolic potential of two symbiotic organisms can be investigated using a merged reaction network generated from reaction networks loaded from contigs databases for the two organisms; the method, `anvio.reactionnetwork.GenomicNetwork.merge_network`, merges the individual networks.

## Chain search

### Consumption and production

A walker can find chains of reactions that produce or consume compounds in the pathway. By default, the `compound_fate` attribute of the walker is set to a value of "consume", but it can also be set to "produce" or "both." "Consumption chains" contain compounds linked by reactions starting with the first compound that is consumed in the chain and proceeding through subsequently consumed compounds to a final product within the pathway map that can no longer be reacted. "Production chains", found when `compound_fate` has a value of "produce", start with the final product that is produced in a chain of reactions and proceed backwards to the first substrate within the pathway map that is not the product of a reaction. Given a `compound_fate` value of "both", both consumption and production chains will be found.

### Target compounds

The `get_chains` method returns all possible chains from a pathway map by default, be they consumption chains, production chains, or both consumption and production chains, given the value of `compound_fate`. If a reaction network is available to the walker, chains are sought starting from compounds in the network. Specific compounds can be targeted using the arguments, `keggcpd_ids` and `modelseed_compound_ids`, e.g., glycine and D-glucose have the the KEGG compound IDs "C00037" and "C00031" and the ModelSEED compound IDs "cpd00033" and "cpd00027", respectively. Note that KEGG compound database IDs differ from the IDs of entries for these compounds in KGML pathway files, which are integer assignments specific to files.

### Reverse chains

A `compound_fate` value of "both" causes `get_chains` to return overlapping consumption and production chains. In the case of a series with one or more irreversible reactions, the method returns one consumption chain starting with the first consumed compound and one production chain starting with the final produced compound. A series of all reversible reactions results in four chains of the same compounds being returned, with the first consumed compound of one chain with the reactions running "forward" being the last produced compound of another chain, and the last produced compound of one chain with the reactions running "reverse" being the first consumed compound of another chain. KGML files are especially useful for the refined analysis of biochemical pathways due to the inclusion of key nonenzymatic reactions and curated reaction reversibilities.

### Max reactions

Certain attributes of a walker constrain the chains returned by `get_chains`. The `max_reactions` attribute truncates chains at the given number of reactions, e.g., a value of 10 causes `get_chains` to report chains of 10 or fewer reactions. With the default value of `None`, chains can continue to indeterminate length.

### Intermediate chains

The attribute, `keep_intermediate_chains`, allows intermediate chains, or subchains, of longer chains to be reported. Intermediate chains are entirely contained within longer chains. By default, this attribute is `False`, and intermediate chains are ignored after being found. If `True`, a very large number of chains can be reported, since every reaction and sequence of reactions within a chain constitutes an intermediate chain.

### Gaps

The `max_gaps` attribute of the walker allows a number of reactions, up to the given value, to be included in reported chains despite not being found in the reaction network, treating the reactions as genomic "gaps." Note that this attribute only applies if the walker has an associated network.

With the attribute, `allow_terminal_gaps`, set to the default value of `False`, gap reactions cannot occur at the beginning or end of a chain; gaps must be flanked on both sides by genomically encoded reactions. The attribute, `allow_alternative_reaction_gaps`, also controls the types of gaps allowed in chains. If a chain links two compounds by a reaction in the network, and there are other "parallel" KGML reactions that also link the compounds but are not represented in the network, then treat these alternative reactions as gaps if the attribute has a value of `True`. With the default value of `False`, the walker does not treat parallel reactions that are not found in the network as gaps.

## Chain attributes

### Compounds and reactions

An `anvio.kgmlnetworkops.Chain` object records a chain of compounds linked by reactions occurring in a KGML representation of a KEGG pathway. `kgml_compound_entries` and `kgml_reactions` attributes of the chain are lists of the elements from the KGML file in the proper order, with there being one more compound than reaction. The `anvio.kgml` library loads KGML (XML type) files into memory as `Pathway` objects, including object representations of each XML element in the file, such as `Entry` and `Reaction` elements.

The boolean `is_consumed` attribute of the chain indicates whether the compounds in the chain are consumed or produced by the corresponding reactions. A consumption chain has a value of `True`, indicating that the first through penultimate compounds of the chain are consumed by the reactions, whereas a production chain has a value of `False`, indicating that the first through penultimate compounds are produced by the reactions, with the first compound being the last produced.

### Reaction directionality

KGML Reaction elements have a certain direction in the file, with compounds specified as "substrates" or "products," and the reactions are also either "reversible" or "irreversible." Reversible reactions may occur in their given direction or in the opposite direction, with "substrates" being products and "products" being substrates of the reaction as it occurs in the chain. The `kgml_reaction_directions` attribute has items corresponding to the reactions of `kgml_reactions`, with an item being `True` if the reaction runs as given in the KGML file, and `False` if the reaction runs in reverse.

### Reaction network associations

Certain attributes of the chain are relevant if the chain was found in the context of genomic annotations stored in a reaction network. These attributes are empty lists if the chain was recovered from the pathway map without reference to a network.

The `gaps` attribute has items corresponding to reactions in the chain, with `False` indicating that the reaction is genomically encoded by a KO in the network, and `True` indicating that the reaction is not encoded and is thus a gap. Chains returned by a walker have a maximum number of `True` values equal to the `max_gaps` value of the walker.

Multiple attributes store associations between KGML compounds and reactions in the chain and network compounds and reactions. The items of the `aliased_modelseed_compounds` attribute correspond to `kgml_compound_entries`, with each item being a tuple of ModelSEED compounds, instances of `anvio.reactionnetwork.ModelSEEDCompound` in the network, mapping to the KGML compound. The items of the `network_kos` and `aliased_modelseed_reactions` attributes correspond to `kgml_reactions`. Each item of `network_kos` is a tuple of KOs, instances of `anvio.reactionnetwork.KO` in the network, mapping to the KGML reaction. Each item of `aliased_modelseed_reactions` is a tuple of ModelSEED reactions, instances of `anvio.reactionnetwork.ModelSEEDReaction` in the network, mapping to the KGML reaction.

### Consumption and production termini

The ideas of a "consumption terminus" and "production terminus" were devised to characterize the connectedness of a chain's first consumed and last produced compound in the KGML map, and are recorded in the boolean attributes, `is_consumption_terminus` and `is_production_terminus`. Note that the first consumed compound is the first compound in the chain if the attribute, `is_consumed`, is `True`, and the last compound if `is_consumed` is `False`, and the last produced compound is the last compound in the chain if `is_consumed` is `True`, and the first compound if `is_consumed` is `False`. Roughly speaking, consumption and production termini are not interconnected intermediates, so a chain running from a consumption to a production terminus begins and ends toward the edge of the map and is less likely to be an intermediate pathway fragment. Consumption and production termini can be defined in a negative sense. `anvio.kgmlnetworkops.KGMLNetworkWalker.check_kgml_compound_entry_consumption_terminus` will find that a compound is not a consumption terminus if (1) it can be produced by an irreversible reaction on the map, or if (2) (a) there are multiple reactions that consume the compound, (b) any one of these reactions is reversible, and (c) not all of the reactions have the same KGML compound products. Another way of defining test (2) in a positive sense is that a compound can be a consumption terminus if multiple reactions consume it, so long as all of the reactions lead to the same products in the map or, if the reactions lead to different products, none of the reactions are reversible. Likewise, `anvio.kgmlnetworkops.KGMLNetworkWalker.check_kgml_compound_entry_production_terminus` will find that a compound is not a production terminus if (1) it can be consumed by an irreversible reaction on the map, or if (2) (a) there are multiple reactions that produce the compound, (b) any one of these reactions is reversible, and (c) not all of the reactions have the same KGML compound reactants. Again, for the sake of completeness, another way of defining test (2) in a positive sense is that a compound can be a production terminus if multiple reactions produce it, so long as all of the reactions come from the same reactants in the map or, if the reactions come from different reactants, none of the reactions are reversible.

### Consumption and production reversibility ranges

The `consumption_reversibility_range` and `production_reversibility_range` attributes record the numbers of compounds that are first consumed and last produced in the chain that are linked by reversible reactions, indicating which compounds can feed into the chain of reactions or be produced by it. Note that the first consumed compound is the first compound in the chain if the attribute, `is_consumed`, is `True`, and the last compound if `is_consumed` is `False`, and the last produced compound is the last compound in the chain if `is_consumed` is `True`, and the first compound if `is_consumed` is `False`. KGML reaction XML elements are annotated as "reversible" or "irreversible." Reversibility ranges are recorded as a tuple of Pythonic start and stop indices. For example, if the first compound of a consumption chain (`is_consumed` is `True`) is linked to the second compound by a reversible reaction, but the second is linked to the third by an irreversible reaction, then the value of `consumption_reversibility_range` is (0, 2). If the last compound of a consumption chain of six compounds is linked to the second-to-last compound by an irreversible reaction, then the value of `production_reversibility_range` is (5, 6).

### Cycles

Presently, chains are a linear sequence of reactions and terminate at cycles. An uncycled compound can be consumed or produced by a "branching" reaction in the map, such as acetyl-CoA entering the Krebs cycle by combining with oxaloacetate to form citrate using citrate synthase. If a chain involves a cycle via a branching reaction, then the chain 

Non-negative integer if the chain is "partly cyclic," meaning that the chain loops a cycle
        via the entry or exit of an uncycled compound in a "branching" reaction that is part of the
        cycle, such as acetyl-CoA entering the Krebs cycle by combining with oxaloacetate to form
        citrate using citrate synthase. The branching reaction occurs twice in the partly cyclic
        chain, once at the end. The non-negative integer is the index of the first occurrence of the
        branching KGML reaction that also ends the chain. -1 if the chain is not partly cyclic.
