A %(trnaseq-fasta)s is a %(fasta)s file of sequences from a single tRNA-seq sample of split that is suitable to be used by %(anvi-trnaseq)s to create a %(trnaseq-db)s.

Like %(contigs-fasta)s files, this file **is required to have simple deflines**. Take a look at your deflines prior to mapping, and remove anything that is not a digit, an ASCII letter, an underscore, or a dash character. The program %(anvi-script-reformat-fasta)s can do this automatically for you with the flag `--simplify-names`.

We recommend using %(anvi-run-workflow)s to create this file from paired-end tRNA-seq reads. The [trnaseq-workflow](../../workflows/trnaseq/) uses [illumina-utils](https://github.com/merenlab/illumina-utils) to merge FASTQ files that may contain a mixture of fully and partially overlapping reads, which both occur using 100 bp (or shorter) reads containing barcodes due to the length of tRNA. Even with 150 bp reads, there may be pre-tRNA covered by partially but not fully overlapping reads. %(anvi-script-reformat-fasta)s comes after illumina-utils in the workflow.
