This artifact contains various information about the SNVs, SCVs, and SAAVs across a %(profile-db)s that is thoroughly described [on this blogpost](http://merenlab.org/2015/07/20/analyzing-variability/#the-output-matrix).  

This was generated by %(anvi-gen-variability-profile)s, which is also described on [that blogpost](http://merenlab.org/2015/07/20/analyzing-variability/#the-anvio-way). 

{:.notice}
Unsure what SNV, SCV, and SAAVs are or looking for a refresher? You can find that information [on the same blogpost](http://merenlab.org/2015/07/20/analyzing-variability/#an-intro-to-single-nucleotidecodonamino-acid-variation). 

In summary, this contains output matrices for your SNVs, SCVs and SAAVs that annotate them with lots of additional infomration. This way, you can interact with that data by running %(anvi-script-snvs-to-interactive)s or incorporate it into your structural analysis when running %(anvi-3dev)s. 

### What kinds of information? 

#### SNVs

For each of your SNVs, this matrix include their position in the contig and gene, sample, coverage data, the A, C, G, and T counts, the reference and consensus nucleotides, entropy value, and more. 

#### SCVs 

This information will only appear if you requested it when running your earlier analysis. To do this, use the flag `--profile-SCVs` when you run %(anvi-profile)s or %(anvi-merge)s. Then, when running %(anvi-gen-variability-profile)s use the flag `--engine CDN`. 

For each SCVs, this matrix details the position, sample, coverage data, count for each of the 64 codons (AAA, AAC, ..., TTG, TTT), entropy, synonymity, etc. 

#### SAAVs

Like the information about SCVs, this information will only appear if you requested it when running your earlier analysis. To do this, use the flag `--profile-SCVs` when you run %(anvi-profile)s or %(anvi-merge)s. Then, when running %(anvi-gen-variability-profile)s use the flag `--engine AA`. 

For each SCVs, this matrix details the position, sample, coverage data, count for each of the 20 amino acids (as well as the stop codon), entropy, BLOSUM62, etc. 

#### Structural information

If you provided %(anvi-gen-variability-profile)s with a %(structure-db)s, then you'll also have some additional columns to your matrices. These include structural annotations, how the residue responds to water, information about bond angles, and a list of residues that are in physical contact with the residue you're looking at. 

For more information on any of this, check out [this page](http://merenlab.org/2015/07/20/analyzing-variability/#the-output-matrix), where every column in these matrices is not only listed, but explained. 
