This program converts one or more %(trnaseq-db)ss into a %(contigs-db)s and %(profile-db)s, forming tRNA seed sequences in the process.

After %(anvi-trnaseq)s is run on a set of samples (or sample splits) from an experiment, the resulting sample databases are analyzed by this program. The %(contigs-db)s and %(profile-db)s generated by this program can be displayed and analyzed like other 'omics data in anvi'o.

As with %(anvi-trnaseq)s, we encourage you to read through the options in the `anvi-convert-trnaseq-database --help` menu, since the various analyses occurring within this program can be manipulated by the user. Most of the heavy lifting occurred in %(anvi-trnaseq)s, meaning that `anvi-convert-trnaseq-database` can easily be run multiple times in a matter of minutes to find, for example, the appropriate number of tRNA seed sequences to display (by adjusting `--max-reported-trna-seeds`) and the necessary coverage cutoffs to isolate positions with modified nucleotides from unfiltered single nucleotide variants (by adjusting `--min-third-fourth-nt`).


### Convert two samples

{{ codestart }}
anvi-convert-trnaseq-database example_trnaseq_database_1 example_trnaseq_database_2 \
                              -o example_empty_output_directory_path
                              -n example_project_name
{{ codestop }}

### Convert two sample splits, assigning priority to the demethylase split in calling the underlying nucleotide at modified positions

{{ codestart }}
anvi-convert-trnaseq-database example_untreated_trnaseq_database example_demethylase_trnaseq_database \
                              -o example_empty_output_directory_path
                              -n example_project_name
                              --preferred-treatment demethylase
{{ codestop }}
